omegaconf/base.py

```diff
@@ -377,28 +400,63 @@ def _resolve_interpolation(
         elif value_kind == ValueKind.STR_INTERPOLATION:
             value = _get_value(value)
             assert isinstance(value, str)
-            orig = value
-            new = ""
-            last_index = 0
-            for match in match_list:
-                new_val = self._resolve_simple_interpolation(
+            try:
+                # NB: `match_list` is ignored here as complex interpolations may be
+                # nested, the string will be re-parsed within `_evaluate_complex()`.
+                new = self._evaluate_complex(
+                    value,
                     key=key,
-                    inter_type=match.group(1),
-                    inter_key=match.group(2),
                     throw_on_missing=throw_on_missing,
                     throw_on_resolution_failure=throw_on_resolution_failure,
                 )
-                # if failed to resolve, return None for the whole thing.
-                if new_val is None:
-                    return None
-                new += orig[last_index : match.start(0)] + str(new_val)
-                last_index = match.end(0)
-
-            new += orig[last_index:]
+            except ParseError:
+                return None
             return StringNode(value=new, key=key)
         else:
             assert False
 
+    def _evaluate_simple(self, value: str, **kw: Any) -> Optional["Node"]:
+        """Evaluate a simple interpolation"""
+        value_kind, match_list = get_value_kind(value=value, return_match_list=True)
+        assert value_kind == ValueKind.INTERPOLATION and len(match_list) == 1
+        match = match_list[0]
+        node = self._resolve_simple_interpolation(
+            inter_type=match.group(1), inter_key=match.group(2), **kw,
+        )
+        return node
+
+    def _evaluate_complex(self, value: str, **kw: Any) -> Optional[str]:
+        """Evaluate a complex interpolation (>1, nested, with other strings...)"""
```

Comment:
Please put a high level description of what this supports and how it works.
In the next major version I am planning to introduce a proper parser generator which will simplify this greatly.
It will help knowing what this function is supposed to be doing when integrating it.

Reply:
Done in ec534f9